x-pack/solutions/security/plugins/security_solution/public/siem_migrations/common/service/migrations_service.ts

```diff
@@ -0,0 +1,210 @@
+  `${DEFAULT_ASSISTANT_NAMESPACE}.${TRACE_OPTIONS_SESSION_STORAGE_KEY}` as const;
+
+export const TASK_STATS_POLLING_SLEEP_SECONDS = 10 as const;
+export const START_STOP_POLLING_SLEEP_SECONDS = 1 as const;
+
+export abstract class SiemMigrationsServiceBase<T extends MigrationTaskStats> {
+  protected abstract startMigrationFromStats(connectorId: string, taskStats: T): Promise<void>;
+  protected abstract fetchMigrationStats(params: GetMigrationStatsParams): Promise<T>;
+  protected abstract fetchMigrationsStatsAll(params: GetMigrationsStatsAllParams): Promise<T[]>;
+  protected abstract sendFinishedMigrationNotification(taskStats: T): void;
+
+  private readonly latestStats$: BehaviorSubject<T[] | null>;
+  private isPolling = false;
+  public connectorIdStorage: MigrationsStorage<string>;
+  public traceOptionsStorage: MigrationsStorage<TraceOptions>;
+
+  constructor(
+    private readonly source: string,
+    protected readonly core: CoreStart,
+    private readonly plugins: StartPluginsDependencies
+  ) {
+    this.connectorIdStorage = new MigrationsStorage<string>(this.source, 'connectorId');
+    this.traceOptionsStorage = new MigrationsStorage<TraceOptions>(this.source, 'traceOptions', {
+      customKey: NAMESPACE_TRACE_OPTIONS_SESSION_STORAGE_KEY,
+      storageType: 'session',
+    });
+
+    this.latestStats$ = new BehaviorSubject<T[] | null>(null);
+
+    this.plugins.spaces.getActiveSpace().then((space) => {
+      this.connectorIdStorage.setSpaceId(space.id);
+      this.startPolling();
+    });
+  }
+
+  /** Returns the latest stats observable, which is updated every time the stats are fetched */
+  public getLatestStats$(): Observable<T[] | null> {
+    return this.latestStats$.asObservable().pipe(distinctUntilChanged(isEqual));
+  }
+
+  /** Returns any missing capabilities for the user to use this feature */
+  public getMissingCapabilities(level?: CapabilitiesLevel): MissingCapability[] {
+    return getMissingCapabilities(this.core.application.capabilities, level);
+  }
+
+  /** Checks if the user has any missing capabilities for this feature */
+  public hasMissingCapabilities(level?: CapabilitiesLevel): boolean {
+    return this.getMissingCapabilities(level).length > 0;
+  }
+
+  /** Checks if the service is available based on the `license`, `capabilities` and `experimentalFeatures` */
+  public isAvailable() {
+    return (
+      !ExperimentalFeaturesService.get().siemMigrationsDisabled &&
+      licenseService.isEnterprise() &&
+      !this.hasMissingCapabilities('minimum')
+    );
+  }
+
+  /** Starts polling the migrations stats if not already polling and if the feature is available to the user */
+  public startPolling() {
+    if (this.isPolling || !this.isAvailable()) {
+      return;
+    }
+    this.isPolling = true;
+    this.startTaskStatsPolling()
+      .catch((e) => {
+        this.core.notifications.toasts.addError(e, {
+          title: i18n.POLLING_ERROR(this.source),
```

Comment:
If the `source` is only used here, I would consider dropping it from the constructor and having a generic title (like `Error fetching migrations stats`) since the error trace will already provide details about it.

Otherwise, if you prefer to have a specific title, we can also abstract it.
```ts
protected abstract getTaskStatsPollingErrorTitle(): string;
```